File: Infrastructure/Repositories/LocationRepository.cs
using Core.Entities.Location;
using Core.Interfaces.Repositories;
using Infrastructure.Data;
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Infrastructure.Repositories
{
    public class LocationRepository : ILocationRepository
    {
        private readonly AppDbContext _db;
        public LocationRepository(AppDbContext db) => _db = db;

        public async Task<IEnumerable<District>> GetAllDistrictsAsync() =>
            await _db.Districts.Where(x => x.IsActive).OrderBy(x => x.Name).ToListAsync();

        public async Task<District?> GetDistrictByIdAsync(int id) =>
            await _db.Districts.FirstOrDefaultAsync(x => x.Id == id && x.IsActive);

        public async Task<IEnumerable<Constituency>> GetConstituenciesByDistrictAsync(int districtId) =>
            await _db.Constituencies
                .Where(x => x.DistrictId == districtId && x.IsActive)
                .OrderBy(x => x.Name)
                .ToListAsync();

        public async Task<IEnumerable<Area>> GetAreasByConstituencyAsync(int constituencyId) =>
            await _db.Areas
                .Where(x => x.ConstituencyId == constituencyId && x.IsActive)
                .OrderBy(x => x.Name)
                .ToListAsync();

        public async Task<IEnumerable<Block>> GetBlocksByAreaAsync(int areaId) =>
            await _db.Blocks
                .Where(x => x.AreaId == areaId && x.IsActive)
                .OrderBy(x => x.Name)
                .ToListAsync();

        public async Task<Block?> GetBlockByIdAsync(int id) =>
            await _db.Blocks.FirstOrDefaultAsync(x => x.Id == id && x.IsActive);
    }
}

File: citizen_connect_python/app/repositories/location_repository.py
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from app.models.location import District, Constituency, Area, Block


class LocationRepository:
    def __init__(self, db: AsyncSession):
        self.db = db

    async def get_all_districts(self) -> list[District]:
        result = await self.db.execute(
            select(District)
            .where(District.is_active == True)
            .order_by(District.name)
        )
        return list(result.scalars().all())

    async def get_constituencies_by_district(
        self, district_id: int
    ) -> list[Constituency]:
        result = await self.db.execute(
            select(Constituency)
            .where(
                Constituency.district_id == district_id,
                Constituency.is_active == True
            )
            .order_by(Constituency.name)
        )
        return list(result.scalars().all())

    async def get_areas_by_constituency(
        self, constituency_id: int
    ) -> list[Area]:
        result = await self.db.execute(
            select(Area)
            .where(
                Area.constituency_id == constituency_id,
                Area.is_active == True
            )
            .order_by(Area.name)
        )
        return list(result.scalars().all())

    async def get_blocks_by_area(self, area_id: int) -> list[Block]:
        result = await self.db.execute(
            select(Block)
            .where(
                Block.area_id == area_id,
                Block.is_active == True
            )
            .order_by(Block.name)
        )
        return list(result.scalars().all())

    async def get_block_by_id(self, block_id: int) -> Block | None:
        result = await self.db.execute(
            select(Block).where(Block.id == block_id, Block.is_active == True)
        )
        return result.scalar_one_or_none()